src/modules/types.rs

```diff
@@ -1,26 +1,108 @@
 use std::fmt::Display;
 
 use heraclitus_compiler::prelude::*;
+use itertools::Itertools;
 use crate::utils::ParserMetadata;
 
-#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[derive(Debug, Clone, Eq, Default)]
 pub enum Type {
     #[default] Null,
     Text,
     Bool,
     Num,
+    Union(Vec<Type>),
     Array(Box<Type>),
     Failable(Box<Type>),
     Generic
 }
 
+impl Type {
+    fn eq_union_normal(one: &[Type], other: &Type) -> bool {
+        one.iter().any(|x| (*x).to_string() == other.to_string())
+    }
+
+    fn eq_unions(one: &[Type], other: &[Type]) -> bool {
+        let bigger;
+        let smaller = if one.len() < other.len() { bigger = other; one } else { bigger = one; other };
+
+        smaller.iter().all(|x| {
+            Self::eq_union_normal(bigger, x)
+        })
+    }
```

Comment:
Those types will not really be equal. Look at this example:
```
fun foo(): Num | [Num] {
  return [1]
}

fun bar(x: Num) {
  return x + 1
}

echo bar(foo())
```

This should raise an error. Just because value can be of some type, we cannot say that it is. We lose type safety here

Reply:
imo union types should not be automatically equal to normal type, but a normal type may be equal to a union type if the latter has the same type in its list.

like

```
Text | Num  !=  Text
Text        ==  Text | Num
Text        !=  Null | Num
```

and to convert `Text | Num` to `Text`, one would check it with an `if x is Text` which would convert `x` to `Text` inside that `if` block

Reply:
Yes, but I don't know if you can check it this way. It's more of a one way casting ability

Reply:
typescript does it that way (mostly)

Reply:
Mostly is crucial here. We shouldn't be able to say that `Text == Text | Num` or `Text | Num == Text`

I know that thanks to this, you're able to do this:
```
fun foo(x: Text | Num) ...

foo(5)
```

But it creates an issue that I've shown before. We need to tackle this issue by checking if a value given to a var/param is the same as declared type or is in the Union type. Not by equating `Text` type with a union type that has `Text` in it.

You need to modify for example how `run_function_with_args` works

Reply:
I agree with @KrosFire. @b1ek, please review [this line of code](https://github.com/amber-lang/amber/blob/1810a90e693a25522d1dc29564d1735fd688ead5/src/modules/function/invocation_utils.rs#L43). Instead of `==` (there actually is `!=` but for simplicity I'll use equal sign) there should be a subset compare function that for union checks if the value is part of the union and for the regular values we can apply just `==` operator as normal.

## Explanation

When it comes to types - we are working on sets, not values. For primitive types it's easy as they are singletons:
```
 Text    Num    Bool
{Text}, {Num}, {Bool}
```
Here we can simply `{Text} == {Text}` because that indeed is the case.

When it comes to the more complex types like unions:
```
 Text | Num   Bool | [Text]
{Text, Num}, {Bool, [Text]} 
```
We need to compare if one set is a subset of another set and for that we cannot rely on Rust's `PartialEq`. This requires a comparison `<=`. So in the line that I've linked above - there should actually be `<=` because we want to check if this is the same value OR is it a subset of this value. This is a not an `eq`. It's an `leq` aka `subset`.